File: problems-originals/1801-1900/1878_get-biggest-three-rhombus-sums-in-a-grid/solution.go
import (
	"sort"
)

func getBiggestThree(grid [][]int) []int64 {
	// Enumerate every (center, k) rhombus by walking its four edges;
	// keep distinct sums and return the three largest.
	m := len(grid)
	n := len(grid[0])
	sums := map[int64]bool{}
	for r := 0; r < m; r++ {
		for c := 0; c < n; c++ {
			for k := 0; ; k++ {
				if r-k < 0 || r+k >= m || c-k < 0 || c+k >= n {
					break
				}
				var total int64
				if k == 0 {
					total = int64(grid[r][c])
				} else {
					for i := 0; i < k; i++ {
						total += int64(grid[r-k+i][c-i])
						total += int64(grid[r+i][c-k+i])
						total += int64(grid[r+k-i][c+i])
						total += int64(grid[r-i][c+k-i])
					}
				}
				sums[total] = true
			}
		}
	}
	out := make([]int64, 0, len(sums))
	for s := range sums {
		out = append(out, s)
	}
	sort.Slice(out, func(i, j int) bool { return out[i] > out[j] })
	if len(out) > 3 {
		out = out[:3]
	}
	return out
}

File: problems-originals/1801-1900/1878_get-biggest-three-rhombus-sums-in-a-grid/solution.java
import java.util.*;

class Solution {

    // Enumerate every (center, k) rhombus by walking its four edges;
    // keep distinct sums and return the three largest.
    public long[] getBiggestThree(int[][] grid) {
        int m = grid.length;
        int n = grid[0].length;
        TreeSet<Long> sums = new TreeSet<>();
        for (int r = 0; r < m; r++) {
            for (int c = 0; c < n; c++) {
                for (int k = 0; ; k++) {
                    if (r - k < 0 || r + k >= m || c - k < 0 || c + k >= n) {
                        break;
                    }
                    long total;
                    if (k == 0) {
                        total = grid[r][c];
                    } else {
                        total = 0;
                        for (int i = 0; i < k; i++) {
                            total += grid[r - k + i][c - i];
                            total += grid[r + i][c - k + i];
                            total += grid[r + k - i][c + i];
                            total += grid[r - i][c + k - i];
                        }
                    }
                    sums.add(total);
                }
            }
        }
        List<Long> desc = new ArrayList<>(sums.descendingSet());
        int take = Math.min(3, desc.size());
        long[] out = new long[take];
        for (int i = 0; i < take; i++) {
            out[i] = desc.get(i);
        }
        return out;
    }
}
